imblearn/over_sampling/_smote_nc.py

```diff
@@ -0,0 +1,246 @@
+    and :ref:`sphx_glr_auto_examples_over-sampling_plot_smote.py`.
+
+    See also
+    --------
+    ADASYN : Over-sample using ADASYN.
+
+    References
+    ----------
+    .. [1] N. V. Chawla, K. W. Bowyer, L. O.Hall, W. P. Kegelmeyer, "SMOTE:
+       synthetic minority over-sampling technique," Journal of artificial
+       intelligence research, 321-357, 2002.
+
+    Examples
+    --------
+
+    >>> from collections import Counter
+    >>> from numpy.random import RandomState
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.preprocessing import OneHotEncoder
+    >>> from imblearn.over_sampling import \
+SMOTE, SMOTENC # doctest: +NORMALIZE_WHITESPACE
+    >>> X, y = make_classification(n_classes=2, class_sep=2,
+    ... weights=[0.1, 0.9], n_informative=3, n_redundant=1, flip_y=0,
+    ... n_features=20, n_clusters_per_class=1, n_samples=1000, random_state=10)
+    >>> print('Original dataset shape (%s, %s)' % X.shape)
+    Original dataset shape (1000, 20)
+    >>> print('Original dataset samples per class {}'.format(Counter(y)))
+    Original dataset samples per class Counter({1: 900, 0: 100})
+    >>> # replace two last columns with nominal features encoded as integers
+    >>> X[:, -2:] = RandomState(10).randint(0, 4, size=(1000, 2))
+    >>> # apply OneHotEncoder
+    >>> encoder = OneHotEncoder(n_values=[4, 4], categorical_features=[18, 19])
+    >>> X = encoder.fit_transform(X)
+    >>> print('One-hot encoded dataset shape (%s, %s)' % X.shape)
+    One-hot encoded dataset shape (1000, 26)
+    >>> sm = SMOTENC(random_state=42, categorical_features=encoder.feature_indices_)
+    >>> X_res, y_res = sm.fit_resample(X, y)
+    >>> print('Resampled dataset samples per class {}'.format(Counter(y_res)))
+    Resampled dataset samples per class Counter({0: 900, 1: 900})
+
+    """
+
+    def __init__(self,
+                 sampling_strategy='auto',
+                 random_state=None,
+                 k_neighbors=5,
+                 m_neighbors='deprecated',
+                 out_step='deprecated',
+                 kind='deprecated',
+                 svm_estimator='deprecated',
+                 n_jobs=1,
+                 ratio=None,
+                 categorical_features=None):
+        super(SMOTENC, self).__init__(sampling_strategy=sampling_strategy,
+                                      random_state=random_state,
+                                      k_neighbors=k_neighbors,
+                                      m_neighbors=m_neighbors,
+                                      out_step=out_step,
+                                      kind=kind,
+                                      svm_estimator=svm_estimator,
+                                      n_jobs=n_jobs,
+                                      ratio=ratio)
+        self.categorical_features = categorical_features
+
+    def _fit_resample(self, X, y):
+        if self.categorical_features is None:
```

Comment:
I generally agree about it and would rather raise a ValueError here in case of None. However, there's a generic assumption about samplers in this library that they should work when constructed with default values only (i.e. smote_nc = SMOTENC()). This assumption is also present in unit tests (see test_common.py) which gathers all the samplers, initializes with defaults and calls ``fit_resample`` on them.

If we raise an error here in case of None: those tests obviously fail and the aforementioned assumption is no longer true.

Reply:
Actually the assumption here is that SMOTENC works with categorical and numerical data only and that it should work with default parameters on those type of data. So "auto" is a good default.

It is true that the common test will fail but in this case (they are only numerical data), and therefore we will change the `X` specifically in the common tests.